content/binary-api.md

```diff
@@ -0,0 +1,212 @@
+
+This can be used to access `private`/`private[this]` definitions within inline definitions.
+
+Example:
+~~~ scala
+class C {
+  @binaryAPIAccessor private def privateAPI: Int = ...
+  @binaryAPIAccessor def publicAPI: Int = ...
+}
+~~~
+will generate the following bytecode signatures
+~~~ java
+public class C {
+  public C();
+  private int privateAPI();
+  public int publicAPI();
+  public final int C$$inline$privateAPI();
+  public final int C$$inline$publicAPI();
+}
+~~~
+
+Note that the change from `private[this]` to package private, protected or public is a binary compatible change.
+Removing this annotation is a binary incompatible change.
+
+#### Binary API and inlining
+
+If there is a reference to a binary API in an inline method we can use the definition without needing an inline accessor.
+
+Example 3:
+~~~ scala
+class C {
+  @binaryAPI protected def a: Int = ...
+  protected def b: Int = ...
+  inline def foo: Int = a + b
+}
+~~~
+before inlining the compiler will generate the accessors for inlined definitions
+~~~ scala
+class C {
+  @binaryAPI protected def a: Int = ...
+  protected def b: Int = ...
+  final def C$inline$b: Int = ...
+  inline def foo: Int = a + C$inline$b
+}
+~~~
+
+Note that if the inlined member is `a` would be private, we would generate the accessor `C$inline$a`, which happens to be binary compatible with the automatically generated one.
+This is only a tiny mitigation of binary compatibility issues compared with all the different ways accessors can be generated.
+
+### Specification
+
+We must add `binaryAPI` and `binaryAPIAccessor` to the standard library.
+
+```scala
+package scala.annotation
+
+final class binaryAPI extends scala.annotation.StaticAnnotation
+final class binaryAPIAccessor extends scala.annotation.StaticAnnotation
+```
+
+#### `@binaryAPI` annotation
+
+* Only valid on `def`, `val`, `lazy val`, `var`, `object`, and `given`.
+* TASTy will contain references to non-public definitions that are out of scope but `@binaryAPI`. TASTy already allows those references.
+* Annotated definition will be public in the generated bytecode. Definitions should be made public as early as possible in the compiler phases, as this can remove the need to create other accessors. It should be done after we check the accessibility of references.
+
+
+#### `@binaryAPIAccessor` annotation
+
+* Only valid on `def`, `val`, `lazy val`, `var`, `object`, and `given`.
+* An public accessor will be generated for the annotated definition. This accessor will be named `<fullClassName>$$inline$<definitionName>`.
+
+#### Inline
+
+* Inlining will not require the generation of an inline accessor for binary APIs.
+* Inlining will not require the generation of a new inline accessor, it will use the binary API accessors.
+* The user will be warned if a new inline accessor is automatically generated.
+  The message will suggest `@binaryAPI` or `@binaryAPIAccessor` and how to fix potential incompatibilities.
```

Comment:
Will the message suggest using either of them? It seems we should always suggest using the latter, because it would break the ABI of existing code to add the former.

Reply:
Both can break the binary API. 

* `@binaryAPI` migration will always break binary compatibility.
* Only if the is `private`/`private[this]` need `@binaryAPIAccessor`. These can break binary compatibility if the class is final.

Reply:
Code like this
```scala
class A:
  private[foo] val foo: Int = 1
  inline def inlined = foo // error
```
would show an error message like this
```scala
-- Error: inline-unstable-accessors.scala:3:4 ------------------------------------
10 |    valBinaryAPI + // error
   |    ^^^^^^^^^^^^^
   |    Generated unstable inline accessor for value foo defined in class A.
   |
   |    Annotate foo with `@binaryAPI` to make it accessible.
   |
   |    Adding @binaryAPI may break binary compatibility if a previous version of this
   |    library was compiled with Scala 3.0-3.3, Binary compatibility should be checked
   |    using MiMa. To keep binary you can add the following accessor to class A:
   |      @binaryAPI private[A] def foo$A$$inline$foo: Int = this.foo
```

More examples can be found in https://github.com/lampepfl/dotty/pull/16992/files#diff-d12e949663e52759a532bf54a827260b0513390bfcdebddf5ff1c1bd04919aaa